File: solutions/csharp/pig-latin/1/PigLatin.cs
using System;
using System.Text.RegularExpressions;

public static class PigLatin
{
    private static string TranslateWord(string word)
    {
        if (Regex.IsMatch(word, "^[aeiouAEIOU]|^xr|^yt")) {
            return word + "ay";
        }
        Match rule3 = Regex.Match(word, "^[^aeiouAEIOU]*qu");
        if (rule3.Success) {
            string prefix = rule3.Value;
            string rest = word.Substring(prefix.Length);
            return rest + prefix + "ay";
        }
        Match rule4 = Regex.Match(word, "^([^aeiouAEIOU]+)y");
        if (rule4.Success) {
            string prefix = rule4.Groups[1].Value;
            string rest = word.Substring(prefix.Length);
            return rest + prefix + "ay";
        }
        Match rule2 = Regex.Match(word, "^[^aeiouAEIOU]+");
        if (rule2.Success) {
            string prefix = rule2.Value;
            string rest = word.Substring(prefix.Length);
            return rest + prefix + "ay";
        }
        return word + "ay";        
    }

    public static string Translate(string phrase)
    {
        string[] words = phrase.Split(' ');
        string result = "";
        foreach (string word in words) {
            result += TranslateWord(word) + " ";
        }
        return result.TrimEnd();
    }
}

File: solutions/typescript/pig-latin/1/pig-latin.ts
function translateWord(word: string) {
    if (word.match(/^[aeiouAEIOU]|^xr|^yt/)) {
        return `${word}ay`;
    }
    const rule3 = word.match(/^[^aeiouAEIOU]*qu/);
    if (rule3) {
        const prefix = rule3[0];
        const rest = word.substring(prefix.length);
        return `${rest}${prefix}ay`;
    }
    const rule4 = word.match(/^([^aeiouAEIOU]+)y/);
    if (rule4) {
        const prefix = rule4[1];
        const rest = word.substring(prefix.length);
        return `${rest}${prefix}ay`;
    }
    const rule2 = word.match(/^[^aeiouAEIOU]+/);
    if (rule2) {
        const prefix = rule2[0];
        const rest = word.substring(prefix.length);
        return `${rest}${prefix}ay`;
    }
    return `${word}ay`;
}

export function translate(phrase: string) {
    const words = phrase.split(' ');
    return words.map(translateWord).join(' ');
}
